File: server/app/services/chat/context/retrieval.py
"""
Context retrieval module for gathering conversation context.

This module handles the retrieval of context components needed for
LLM processing:
- Recent messages from the session
- RAG-based relevant segments from vector store
- Latest summary from the summary system

The retrieval process supports both RAG-enabled and RAG-disabled modes.
"""

from sqlalchemy.orm import Session
from typing import List, Dict, Any, Optional
from app.models.message import Message
from app.models.embedding_config import EmbeddingConfig
from app.services.chat.context.summary import SummaryService
from app.services.chat.context.vector_store import VectorStoreService
from app.services.data_service import DataService
from app.logger import logger


class RetrievalService:
    """
    Service for retrieving context components for chat processing.
    
    This service coordinates the retrieval of:
    - Recent messages (with optional status filter)
    - RAG segments (if embedding is configured)
    - Latest summary (if available)
    """
    
    @staticmethod
    def get_embedding_config_for_session(db: Session, session_id: int) -> Optional[EmbeddingConfig]:
        """
        Get the embedding configuration for a session's agent.
        
        Args:
            db: Database session
            session_id: Session ID
            
        Returns:
            EmbeddingConfig if configured, None otherwise
        """
        session = DataService.get_session(db, session_id)
        if not session:
            return None

        agent = DataService.get_agent(db, session.agent_id)
        if not agent:
            return None

        # Check if agent has embedding config (ID > 0 means configured)
        if agent.embedding_config_id and agent.embedding_config_id > 0:
            return db.query(EmbeddingConfig).filter(
                EmbeddingConfig.id == agent.embedding_config_id
            ).first()

        return None

    @staticmethod
    def get_recent_messages(
        db: Session,
        session_id: int,
        limit: int = 10,
        status: Optional[int] = None
    ) -> List[Dict[str, Any]]:
        """
        Get recent messages from a session.
        
        Messages are returned in chronological order (oldest first).
        Optionally filter by message status.
        
        Args:
            db: Database session
            session_id: Session ID
            limit: Maximum number of messages to retrieve
            status: Optional message status filter
            
        Returns:
            List of message dictionaries with 'role', 'content', and 'id' keys
        """
        query = db.query(Message).filter(
            Message.session_id == session_id
        )
        
        # Apply status filter if provided
        if status is not None:
            query = query.filter(Message.status == status)
        
        # Get messages in reverse order (newest first), then reverse
        messages = query.order_by(Message.id.desc()).limit(limit).all()
        messages = list(reversed(messages))

        return [
            {"role": msg.role, "content": msg.content, "id": msg.id}
            for msg in messages
        ]

    @staticmethod
    def get_context_without_rag(
        db: Session,
        session_id: int,
        recent_count: int = 10
    ) -> Dict[str, Any]:
        """
        Get context without RAG retrieval.
        
        Used for queries that don't need RAG (e.g., greetings, chitchat).
        Only retrieves recent messages and latest summary.
        
        Args:
            db: Database session
            session_id: Session ID
            recent_count: Number of recent messages to retrieve
            
        Returns:
            Dictionary with 'recent_messages', 'relevant_segments', and 'summary'
        """
        result = {
            "recent_messages": [],
            "relevant_segments": [],
            "summary": None
        }

        from app.models.message import MESSAGE_STATUS_COMPLETED
        
        # Get recent completed messages
        result["recent_messages"] = RetrievalService.get_recent_messages(
            db, session_id, recent_count, status=MESSAGE_STATUS_COMPLETED
        )

        # Get latest summary (may be older than current message count)
        latest_summary = SummaryService.get_latest_summary(db, session_id)
        if latest_summary:
            result["summary"] = {
                "version": latest_summary.version,
                "content": latest_summary.content
            }

        return result

    @staticmethod
    def get_context_for_chat(
        db: Session,
        session_id: int,
        query: str,
        recent_count: int = 10,
        rag_count: int = 5
    ) -> Dict[str, Any]:
        """
        Get full context for chat processing with RAG.
        
        This method retrieves all context components:
        1. Recent messages from the session
        2. RAG segments relevant to the query (if embedding configured)
        3. Latest summary (if available)
        
        Args:
            db: Database session
            session_id: Session ID
            query: Processed query for RAG search
            recent_count: Number of recent messages to retrieve
            rag_count: Number of RAG segments to retrieve
            
        Returns:
            Dictionary with 'recent_messages', 'relevant_segments', 
            'summary', and 'has_embedding' keys
        """
        result = {
            "recent_messages": [],
            "relevant_segments": [],
            "summary": None,
            "has_embedding": False
        }

        from app.models.message import MESSAGE_STATUS_COMPLETED
        
        # Get recent completed messages
        result["recent_messages"] = RetrievalService.get_recent_messages(
            db, session_id, recent_count, status=MESSAGE_STATUS_COMPLETED
        )

        # Perform RAG retrieval if embedding is configured
        embedding_config = RetrievalService.get_embedding_config_for_session(db, session_id)
        if embedding_config:
            result["has_embedding"] = True
            try:
                vector_store = VectorStoreService.get_instance(session_id, embedding_config)

                # Search for relevant segments
                search_results = vector_store.search(query, k=rag_count)
                result["relevant_segments"] = search_results
                logger.info(f"RAG retrieved {len(search_results)} segments for session {session_id}")
            except Exception as e:
                logger.error(f"RAG retrieval failed: {str(e)}", exc_info=True)

        # Get latest summary (may be older than current message count)
        latest_summary = SummaryService.get_latest_summary(db, session_id)
        if latest_summary:
            result["summary"] = {
                "version": latest_summary.version,
                "content": latest_summary.content
            }

        return result

    @staticmethod
    def index_messages(
        db: Session,
        session_id: int,
        message_ids: List[int]
    ) -> bool:
        """
        Index messages in the vector store for RAG retrieval.
        
        This method adds messages to the vector store after they are
        completed, enabling future RAG retrieval.
        
        Args:
            db: Database session
            session_id: Session ID
            message_ids: List of message IDs to index
            
        Returns:
            True if indexing succeeded, False otherwise
        """
        # Check if embedding is configured
        embedding_config = RetrievalService.get_embedding_config_for_session(db, session_id)
        if not embedding_config:
            logger.info(f"No embedding config for session {session_id}, skipping indexing")
            return False

        # Load messages to index
        messages = db.query(Message).filter(
            Message.id.in_(message_ids),
            Message.session_id == session_id
        ).all()

        if not messages:
            logger.warning(f"No messages found for indexing in session {session_id}")
            return False

        try:
            vector_store = VectorStoreService.get_instance(session_id, embedding_config)

            # Prepare data for indexing
            contents = [msg.content for msg in messages]
            metadatas = [
                {"role": msg.role, "message_id": msg.id}
                for msg in messages
            ]

            # Add to vector store
            vector_store.add_messages(
                message_ids=[msg.id for msg in messages],
                contents=contents,
                metadatas=metadatas
            )

            logger.info(f"Indexed {len(messages)} messages for session {session_id}")
            return True

        except Exception as e:
            logger.error(f"Failed to index messages: {str(e)}", exc_info=True)
            return False

File: server/internal/service/chat/chatctx/retrieval.go
package chatctx

import (
	"private-buddy-server/internal/database"
	"private-buddy-server/internal/model"
	"private-buddy-server/internal/service/llm"
	"private-buddy-server/internal/service/vectorstore"

	applogger "private-buddy-server/internal/logger"
)

// Segment source constants
const (
	SourceChatHistory = iota + 1
	SourceKnowledgeBase
)

// Segment represents a retrieved context segment used in prompt assembly.
type Segment struct {
	Content string `json:"content"`
	Source  int    `json:"source"`
}

// RetrievalResult holds all context components retrieved for chat processing.
type RetrievalResult struct {
	RecentMessages   []model.Message `json:"recent_messages"`
	RelevantSegments []Segment       `json:"relevant_segments"`
	SummaryVersion   int             `json:"summary_version"`
	Narrative        string          `json:"narrative"`
	HasEmbedding     bool            `json:"has_embedding"`
}

// GetEmbeddingConfigForSession returns the embedding config for a session's agent.
// Traverses session -> agent -> embedding_config to find the configuration.
// Returns nil if any step fails (session not found, agent not found, no config).
func GetEmbeddingConfigForSession(sessionID int64) *model.EmbeddingConfig {
	var session model.Session
	if err := database.DB.First(&session, sessionID).Error; err != nil {
		return nil
	}

	var agent model.Agent
	if err := database.DB.First(&agent, session.AgentID).Error; err != nil {
		return nil
	}

	if agent.EmbeddingConfigID > 0 {
		var config model.EmbeddingConfig
		if err := database.DB.First(&config, agent.EmbeddingConfigID).Error; err != nil {
			return nil
		}
		return &config
	}

	return nil
}

// GetRecentMessages returns recent messages from a session in chronological order.
// Messages are fetched in DESC order by ID and then reversed to ASC order.
// If status >= 0, only messages with that status are returned; -1 means no filter.
func GetRecentMessages(sessionID int64, limit int, status int) []model.Message {
	query := database.DB.Model(&model.Message{}).Where("session_id = ?", sessionID)

	if status >= 0 {
		query = query.Where("status = ?", status)
	}

	var messages []model.Message
	query.Order("id DESC").Limit(limit).Find(&messages)

	for i, j := 0, len(messages)-1; i < j; i, j = i+1, j-1 {
		messages[i], messages[j] = messages[j], messages[i]
	}

	return messages
}

// buildSummaryAndNarrative extracts summary version and cached narrative from a HistoricalSummary.
// Returns (nil, nil) if latestSummary is nil.
func buildSummaryAndNarrative(latestSummary *model.HistoricalSummary) (int, string) {
	if latestSummary == nil {
		return -1, ""
	}

	return latestSummary.Version, latestSummary.Narrative
}

// GetContextWithoutRAG retrieves context without RAG retrieval.
// Used for queries that don't need RAG (e.g., greetings, chitchat).
// Retrieves recent messages, latest summary, and cached narrative.
func GetContextWithoutRAG(sessionID int64, recentCount int) *RetrievalResult {
	result := &RetrievalResult{
		RecentMessages:   []model.Message{},
		RelevantSegments: []Segment{},
	}

	result.RecentMessages = GetRecentMessages(sessionID, recentCount, model.MessageStatusCompleted)

	latestSummary := GetLatestSummaryByID(sessionID)
	result.SummaryVersion, result.Narrative = buildSummaryAndNarrative(latestSummary)

	return result
}

// GetContextForChat retrieves context for chat response generation.
// Returns:
//  1. Recent messages from the session
//  2. RAG segments relevant to the query (if embedding configured)
//  3. Latest summary (if available)
//  4. Cached narrative from summary record (if available)
func GetContextForChat(sessionID int64, query string, recentCount int, ragCount int) *RetrievalResult {
	result := &RetrievalResult{
		RecentMessages:   []model.Message{},
		RelevantSegments: []Segment{},
		HasEmbedding:     false,
	}

	result.RecentMessages = GetRecentMessages(sessionID, recentCount, model.MessageStatusCompleted)

	embeddingConfig := GetEmbeddingConfigForSession(sessionID)
	if embeddingConfig != nil {
		result.HasEmbedding = true
		embeddingSvc := llm.NewEmbeddingService(embeddingConfig.BaseURL, embeddingConfig.APIKey, embeddingConfig.ModelID, 0)
		vectorStore := vectorstore.NewVectorStoreService(embeddingSvc)
		if err := vectorStore.Init(); err == nil {
			searchResults, err := vectorStore.Search(sessionID, query, ragCount)
			if err != nil {
				applogger.L.Error("RAG retrieval failed", "error", err)
			} else {
				for _, sr := range searchResults {
					result.RelevantSegments = append(result.RelevantSegments, Segment{
						Content: sr.Content,
						Source:  SourceChatHistory,
					})
				}
				applogger.L.Info("RAG retrieved segments",
					"session_id", sessionID,
					"count", len(searchResults),
				)
			}
			vectorStore.Close()
		}
	}

	latestSummary := GetLatestSummaryByID(sessionID)
	result.SummaryVersion, result.Narrative = buildSummaryAndNarrative(latestSummary)

	return result
}

// IndexMessages adds messages to the vector store for RAG retrieval.
// This method adds messages to the vector store after they are completed,
// enabling future RAG retrieval. Returns true if indexing succeeded.
//
// NOTE: This only indexes the given messageIDs (typically the current round).
// Messages that existed before embedding was configured are NOT retroactively
// indexed. A batch re-index mechanism is needed to cover that case.
func IndexMessages(sessionID int64, messageIDs []int64) bool {
	embeddingConfig := GetEmbeddingConfigForSession(sessionID)
	if embeddingConfig == nil {
		applogger.L.Info("No embedding config for session, skipping indexing", "session_id", sessionID)
		return false
	}

	var messages []model.Message
	database.DB.Where("id IN ? AND session_id = ?", messageIDs, sessionID).Find(&messages)

	if len(messages) == 0 {
		applogger.L.Warn("No messages found for indexing", "session_id", sessionID)
		return false
	}

	embeddingSvc := llm.NewEmbeddingService(embeddingConfig.BaseURL, embeddingConfig.APIKey, embeddingConfig.ModelID, 0)
	vectorStore := vectorstore.NewVectorStoreService(embeddingSvc)
	if err := vectorStore.Init(); err != nil {
		applogger.L.Error("Failed to init vector store for indexing", "error", err)
		return false
	}
	defer vectorStore.Close()

	contents := make([]string, len(messages))
	metadatas := make([]vectorstore.VectorMetadata, len(messages))
	msgIDs := make([]int64, len(messages))

	for i, msg := range messages {
		contents[i] = msg.Content
		metadatas[i] = vectorstore.VectorMetadata{
			MessageID: msg.ID,
			Role:      msg.Role,
			Content:   msg.Content,
		}
		msgIDs[i] = msg.ID
	}

	if err := vectorStore.AddMessages(sessionID, msgIDs, contents, metadatas); err != nil {
		applogger.L.Error("Failed to index messages", "error", err)
		return false
	}

	applogger.L.Info("Indexed messages for session", "session_id", sessionID, "count", len(messages))
	return true
}
